### PR TITLE
Recursive filter_none in Inference Providers

### PR DESCRIPTION
**Pull request by Wauplin**
`None` values in payload can lead to a BadRequest error if the Inference Provider did not expect the field and is doing input validation. This is for instance what's happening in https://github.com/huggingface/huggingface_hub/pull/3170 with groq complaining if `tool_calls` and `tool_call_id` are both set (with None values in them). Groq is complying to the OpenAI specs, this is more a problem on our side but won't be in this PR (*).

What we currently do to avoid these issues is to filter out None values using `filter_none`. However this helper only removes None values at the top-level of the dictionary. **This PR updates `filter_none` to recursively remove None values.** I added some tests to ensure that values like `""`, `0` and `[]` are not removed. This PR should fix the Groq issue and will likely avoid some other errors in the future. 

---

_(*) regarding why we don't exactly follow the OpenAI specs in our inference types is that we don't have a good way to generate Union type. If a schema is defined as "oneOf", our script will generate a class that is in fact a superset of all possible values, with all of them being optional. See https://github.com/huggingface/huggingface.js/pull/1537 or https://github.com/glideapps/quicktype/issues/1266 for more details. This issue might be annoying on the long run but there is no easy way to address it. Pydantic defines [discriminators unions](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions) for that which we could also implement. What's harder is to have a reliable and generic `parse_obj` method extending a Union type (which doesn't seem possible). Let's see if/when it becomes for urgent to tackle it._

cc @Vaibhavs10 @burtenshaw

### CODE REVIEW

File: src/huggingface_hub/inference/_providers/_common.py [hunk 1]
```diff
@@ -36,8 +37,32 @@
 }
 
 
-def filter_none(d: Dict[str, Any]) -> Dict[str, Any]:
-    return {k: v for k, v in d.items() if v is not None}
+def filter_none(obj: Union[Dict[str, Any], List[Any]]) -> Dict[str, Any]:
```
Comment (hanouticelina) [inline src/huggingface_hub/inference/_providers/_common.py]: @Wauplin could you double check if I didn't miss anything here? 😄

File: src/huggingface_hub/inference/_providers/_common.py [hunk 2]
```diff
@@ -224,9 +249,12 @@ def _prepare_route(self, mapped_model: str, api_key: str) -> str:
...
 
     def _prepare_payload_as_dict(
-        self, inputs: Any, parameters: Dict, provider_mapping_info: InferenceProviderMapping
+        self,
+        inputs: List[Union[Dict, ChatCompletionInputMessage]],
+        parameters: Dict,
+        provider_mapping_info: InferenceProviderMapping,
     ) -> Optional[Dict]:
-        return {"messages": inputs, **filter_none(parameters), "model": provider_mapping_info.provider_id}
+        return filter_none({"messages": inputs, **parameters, "model": provider_mapping_info.provider_id})
```
Comment (hanouticelina) [inline src/huggingface_hub/inference/_providers/_common.py:255]: filtering None values from parameters does not really solve the issue since tool_calls, tool_call_id live inside messages, so we need to filter None values from messages as well

File: tests/test_inference_providers.py [hunk 3]
```diff
@@ -1252,6 +1344,14 @@ def test_recursive_merge(dict1: Dict, dict2: Dict, expected: Dict):
...
             {"a": [0, 1, None]},
         ),
+        # dicts inside list are cleaned, list level None kept
+        ({"a": [{"x": None, "y": 1}, None]}, {"a": [{"y": 1}, None]}),
+        # remove every None that is the value of a dict key
+        (
+            [None, {"x": None, "y": 5}, [None, 6]],
+            [None, {"y": 5}, [None, 6]],
+        ),
+        ({"a": [None, {"x": None}]}, {"a": [None]}),
```
Comment (Wauplin) [inline tests/test_inference_providers.py]:
```suggestion
        ({"a": [None, {"x": None}]}, {"a": [None], {}}),
```

Very much a corner case here but I think it's preferable to keep the same list length in case of empty list. It feels less error-prone since we are always keeping None values in lists.

Comment (hanouticelina) [reply tests/test_inference_providers.py]: makes sense! fixed in [03019ff](https://github.com/huggingface/huggingface_hub/pull/3178/commits/03019ff38730d64e92e8e7447c5f4f5648cb6d2f)

File: src/huggingface_hub/inference/_providers/_common.py [hunk 4]
```diff
@@ -36,8 +37,30 @@
 }
 
 
-def filter_none(d: Dict[str, Any]) -> Dict[str, Any]:
-    return {k: v for k, v in d.items() if v is not None}
+def filter_none(obj: Union[Dict[str, Any], List[Any]]) -> Dict[str, Any]:
```
Comment (Wauplin) [inline src/huggingface_hub/inference/_providers/_common.py]: I think we can be more precise here to tell _"if Dict as input, Dict as output, if List as input, List as output"_

```suggestion
@overload
def filter_none(obj: Dict[str, Any]) -> Dict[str, Any]: ...
@overload
def filter_none(obj: List[Any]) -> List[Any]: ...

def filter_none(obj: Union[Dict[str, Any], List[Any]]]) -> Union[Dict[str, Any], List[Any]]:
```

(+ import `typing.overload`)

File: src/huggingface_hub/inference/_providers/_common.py [hunk 5]
```diff
@@ -36,8 +37,30 @@
...
+        return cleaned
+
+    if isinstance(obj, list):
+        cleaned_list: List[Any] = []
+        for v in obj:
+            if isinstance(v, (dict, list)):
+                v = filter_none(v)
+
+            cleaned_list.append(v)
+        return cleaned_list  # type: ignore [return-value]
```
Comment (Wauplin) [inline src/huggingface_hub/inference/_providers/_common.py]:
```suggestion
    if isinstance(obj, list):
        return [filter_none(v) if isinstance(v, (dict, list)) else v for v in obj]
```

slight preference for a list comprehension here (I feel it's clearer that we guarantee the list length doesn't change)

Also the `#type: ignore` can be removed if type hints is improved above

File: tests/test_inference_providers.py [hunk 6]
```diff
@@ -1152,6 +1153,98 @@ def test_prepare_payload(self):
...
+                        "role": "tool",
+                        "content": "pong",
+                        "tool_call_id": "abc123",
+                        "name": "dummy_tool",
+                    },
+                ],
+            ),
+        ],
+    )
+    def test_prepare_payload_filters_messages(self, raw_messages, expected_messages):
```
Comment (Wauplin) [inline tests/test_inference_providers.py:1222]: thanks for adding this :)

**Review by Wauplin — COMMENTED** [review]
Left some minor comments but the logic itself looks good to me. I can't approve the PR since it's originally mine but LGTM once comments are addressed :heavy_check_mark:

File: src/huggingface_hub/inference/_providers/_common.py (hunk 5 above)
Comment (hanouticelina) [reply src/huggingface_hub/inference/_providers/_common.py]: yes addressed in [aba4227](https://github.com/huggingface/huggingface_hub/pull/3178/commits/aba422721a5795fe1a9917f50411a9794d9ab699)

File: src/huggingface_hub/inference/_providers/_common.py (hunk 4 above)
Comment (hanouticelina) [reply src/huggingface_hub/inference/_providers/_common.py]: addressed in [aba4227](https://github.com/huggingface/huggingface_hub/pull/3178/commits/aba422721a5795fe1a9917f50411a9794d9ab699)
